src/resource/resource_base.hpp

```diff
@@ -14,11 +14,15 @@ class ResourceBase;
 using Dependencies = std::unordered_map<Name, std::shared_ptr<ResourceBase>>;
 class ResourceBase {
    public:
-    ResourceBase(ResourceType type) : type(std::move(type)){};
-
-    std::string name;
-    ResourceType type;
+    ResourceBase(ResourceType type) : type_(std::move(type)){};
+    static ResourceType base_type;
     virtual grpc::StatusCode stop(std::unordered_map<std::string, ProtoType*> extra);
     virtual grpc::StatusCode stop();
     virtual void reconfigure(Dependencies deps, Resource cfg);
+    virtual std::string name();
+    ResourceType type();
```

Comment:
Is there a way to enforce that all subclasses of a `ResourceBase` implement their own `from_robot`?

Reply:
update: from speaking with Drew I think we actually have a good solution for this. It breaks a bit from other SDKs (specifically, it keeps the logic within `robot` rather than the specific resources) but should be cleaner, more idiomatic, and enforce method existence. Stand by :)